### PR TITLE
[supervisor]: gitpod config changed analytics

### PR DESCRIPTION
**Pull request by akosyakov**
## Description

This PR adds analytics for .gitpod.yml changes. We are track when changes are persisted to a disk and debounce them for 5s. After last change we analyse changes to top-level elements and report an event for each. It allows us to gain insights in what users are changing on top-level, e.g `ports` or `tasks`,   and avoid a lot of traffic at the same time. Later we can do more fine grained analysis to see which object properties, like `ports.name`.

I had to resolve couple issues which were discovered during review or were blocking testing:
- enable analytics env inside workspace in preview environments to verify with segment
- don't fire change events for syntactically broken .gitpod.yml files, thank you @andreafalzetti for noticing it
- adding git hook which keeps goland types up to date with gitpod schema, otherwise we will miss new properties, thank you @mustard-mh for brining forward that yaml parser only see properties which are defined in goland

## Related Issue(s)

Fixes https://github.com/gitpod-io/gitpod/issues/6894

## How to test

- Open staging untrusted.
- Start a workspace, open .gitpod.yml yet, wait for 5s.
- You should not yet see any `gitpod_config_changed` events.
- Now edit .gitpoy.yml, then you finish wait for 5s.
- Now you should see `gitpod_config_changed` events.
- Edit .gitpod.yml to add some syntax errors and semantic changes, then you finish wait for 5s.
- You should not yet see any `gitpod_config_changed` events.
- Edit .gitpod.yml to fix syntax errors, then you finish wait for 5s.
- Now you should see `gitpod_config_changed` events for semantic changes.

Check for regressions:
- Add new ports and check there are detected in the ports view.

## Release Notes

```release-note
NONE
```

## Werft options:

- [x] /werft with-preview
- [x] /werft analytics=segment|TEZnsG4QbLSxLfHfNieLYGF4cDwyFWoe

### CODE REVIEW

**Comment by roboquat** [comment]
@akosyakov: Adding the "do-not-merge/release-note-label-needed" label because no release-note block was detected, please follow our [release note process](https://git.k8s.io/community/contributors/guide/release-notes.md) to remove it.

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository.
</details>

**Comment by akosyakov** [comment]
Important if you want to rebuild then you start a Gitpod workspace and do following:
`werft run github -s .werft/jobs/build/installer/installer.ts -s .werft/jobs/build/installer/post-process.sh -f`

Because of security concerns, changes to werft scripts are not picked by default, but scripts from main are used. We need to side load them.

File: components/supervisor/pkg/config/gitpod-config-analytics.go
Comment (andreafalzetti) [inline components/supervisor/pkg/config/gitpod-config-analytics.go:44]:
```suggestion
	a.log.Debug("gitpod config analytics: scheduling")
```

File: components/supervisor/pkg/config/gitpod-config-analytics.go
Comment (akosyakov) [inline components/supervisor/pkg/config/gitpod-config-analytics.go:85]: `gitpod config analytics:`

File: components/supervisor/pkg/config/gitpod-config-analytics.go
Comment (akosyakov) [inline components/supervisor/pkg/config/gitpod-config-analytics.go:112]: `gitpod config analytics:`

File: components/supervisor/pkg/supervisor/supervisor.go
Comment (akosyakov) [inline components/supervisor/pkg/supervisor/supervisor.go:1518]: `gitpod config analytics: failed to fetch workspace owner`

File: components/supervisor/pkg/supervisor/supervisor.go
Comment (andreafalzetti) [inline components/supervisor/pkg/supervisor/supervisor.go:1516]: Is there any benefit on calling the API instead of using `os.Getenv("GITPOD_WORKSPACE_ID")`?

Comment (andreafalzetti) [reply components/supervisor/pkg/supervisor/supervisor.go:1516]: never mind, I just noticed it's for the owner

Comment (akosyakov) [reply components/supervisor/pkg/supervisor/supervisor.go:1516]: yes, GITPOD_WORKSPACE_ID is internal implementation, Config abstracts away from it.

File: components/supervisor/pkg/config/gitpod-config-analytics.go
Comment (mustard-mh) [inline components/supervisor/pkg/config/gitpod-config-analytics.go:82]: This function will always returns all keys of `.gitpod.yml` since they are init-ed as a struct, i.e. pointer->nil string->"" int->0.

Or this is as expected?

Comment (akosyakov) [reply components/supervisor/pkg/config/gitpod-config-analytics.go:82]: I don't think it is true, at least I don't see it while testing. I can add more tests.

File: components/supervisor/pkg/supervisor/supervisor.go
Comment (mustard-mh) [reply components/supervisor/pkg/supervisor/supervisor.go:1516]: Yes, @akosyakov always said that once we use `GITPOD_*` env var [[1](https://github.com/gitpod-io/gitpod/pull/8313#discussion_r810007160)] 😂 

But I'm not sure if we have a chance to deprecate it since many users are using Gitpod's env vars. So maybe it's safe somehow? Anyway it must be best if we don't use it

File: components/supervisor/pkg/config/gitpod-config-analytics.go
Comment (mustard-mh) [reply components/supervisor/pkg/config/gitpod-config-analytics.go:82]: This is from the first test case

image

File: components/supervisor/pkg/supervisor/supervisor.go
Comment (akosyakov) [reply components/supervisor/pkg/supervisor/supervisor.go:1516]: there is somethere an issue to provide `gp info` instead and remove all `GITPOD_*` from docs

**Comment by mustard-mh** [comment]
We only get the diffs for the first layer of fields, is it enough? Maybe `inner-loop` can re-use this `ConfigAnalyzer` but it required more detail of diff (i.e. tasks.?.name don't need to trigger, but tasks.?.command need)

**Comment by andreafalzetti** [comment]
When the YAML file is invalid, additional events are fired, loom: https://www.loom.com/share/e140a1decc6c4019bbb41785ec280c72

**Comment by mustard-mh** [comment]
> When the YAML file is invalid, additional events are fired, loom: https://www.loom.com/share/e140a1decc6c4019bbb41785ec280c72

Yes, it treated this object as nil 🙈, so diff from the previous one, will return all previous fields.

**Comment by akosyakov** [comment]
> We only get the diffs for the first layer of fields, is it enough? Maybe inner-loop can re-use this ConfigAnalyzer but it required more detail of diff (i.e. tasks.?.name don't need to trigger, but tasks.?.command need)

It is an idea to have deep structural analysis, but with another PR.

**Review by sagor999 — APPROVED** [review]
reviewed one file that workspace team is code owner of and it LGTM :smile:

**Review by geropl — APPROVED** [review]

**Comment by akosyakov** [comment]
/hold

because it does not handle invalid files well and it turned out we have a bug generally in gitpod config watching, i.e. fire nil config for invalid

File: components/gitpod-protocol/go/gitpod-config-types.go
Comment (akosyakov) [inline components/gitpod-protocol/go/gitpod-config-types.go:1]: This file auto-generated. I added git hook which regenerates it on commit, see [.pre-commit-config.yaml](https://github.com/gitpod-io/gitpod/pull/11653/files#diff-63a9c44a44acf85fea213a857769990937107cf072831e1a26808cfde9d096b9)

File: components/gitpod-protocol/data/gitpod-schema.json
Comment (akosyakov) [inline components/gitpod-protocol/data/gitpod-schema.json:264]: Refactored to extract a common properties to the definition and stabilize code generation.

**Comment by felladrin** [comment]
ℹ️ Reviewing it now.

File: components/supervisor/pkg/supervisor/supervisor.go
Comment (felladrin) [inline components/supervisor/pkg/supervisor/supervisor.go:1540]: **Observation (non-blocking)**: Opening a workspace without .gitpod.yml and then running `gp init` triggered 6 `gitpod_config_changed` at once, one for each of the keys:
- MainConfiguration
- WorkspaceLocation
- Ports
- CheckoutLocation
- Tasks
- ExperimentalNetwork

image

The content of the generated .gitpod.yml was the following:

```yml
# List the start up tasks. Learn more https://www.gitpod.io/docs/config-start-tasks/
tasks:
  - init: echo 'init script' # runs during prebuild
    command: echo 'start script'

# List the ports to expose. Learn more https://www.gitpod.io/docs/config-ports/
ports:
  - port: 3000
    onOpen: open-preview
```

So I expected it would trigger just two events keys: "Tasks" and "Ports".

**Review by felladrin — APPROVED** [review]
Working as advertised 🎉 

I left just a comment about events triggered when the workspace has no `.gitpod.yml` and we run `gp init`, which is an edge case.

File: components/supervisor/pkg/supervisor/supervisor.go
Comment (akosyakov) [reply components/supervisor/pkg/supervisor/supervisor.go:1540]: rrr, it is a bug, let me have a look a bit later, what is going on

**Comment by akosyakov** [comment]
> I left just a comment about events triggered when the workspace has no .gitpod.yml and we run gp init, which is an edge case.

let me verify first that it does not happen then a user creates an empty file as well, it does not look good

**Comment by akosyakov** [comment]
@felladrin I pushed the fix when https://werft.gitpod-dev.com/job/gitpod-build-ak-identiy-config-changed.38 is ready you can try again.

**Review by andreafalzetti — APPROVED** [review]
Great work @akosyakov 🙏 

It works as expected when the `.gitpod.yml` is invalid. I've noticed that when creating or deleting the whole file it does detect changes for the entire file but I'd say it's kind of expected and I don't consider it a bug.

**Comment by akosyakov** [comment]
@loujaybee @andreafalzetti Do you think it would be useful to add an attribute like `operation` to distnguish what happened, i.e. `added`, `updated`,  `removed`? I looked right now at analytics it can be misleading, i.e. you can see 2 modifications to tasks, but final maybe was removing so actually tasks are not used. For analytics it maybe interesting to focus only on `added` and `updated`? It is rather an easy fix.

**Comment by akosyakov** [comment]
> @loujaybee @andreafalzetti Do you think it would be useful to add an attribute like operation to distinguish what happened, i.e. added, updated, removed? I looked right now at analytics it can be misleading, i.e. you can see 2 modifications to tasks, but final maybe was removing so actually tasks are not used. For analytics it maybe interesting to focus only on added and updated? It is rather an easy fix.

Actually I'm wrong we cannot distinguish `removed` from `updated` because our goland types using string types not pointers. It will be quite impactful change, let's leave it for later.

**Pull request merged**
